File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/connectors/google/GoogleHttpConnector.java
package ru.kmorozov.gbd.core.logic.connectors.google;

import com.google.api.client.http.*;
import com.google.api.client.http.javanet.NetHttpTransport;
import ru.kmorozov.gbd.core.config.GBDOptions;
import ru.kmorozov.gbd.core.logic.Proxy.HttpHostExt;
import ru.kmorozov.gbd.core.logic.connectors.HttpConnector;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.net.URI;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Created by km on 17.05.2016.
 */
public class GoogleHttpConnector extends HttpConnector {

    private static final Map<String, HttpRequestFactory> httpFactoryMap = new ConcurrentHashMap<>();

    private HttpRequestFactory getFactory(HttpHostExt proxy) {
        String key = HttpConnector.getProxyKey(proxy);

        HttpRequestFactory requestFactory = GoogleHttpConnector.httpFactoryMap.get(key);

        if (null == requestFactory) synchronized (proxy) {
            requestFactory = new NetHttpTransport.Builder().setProxy(proxy.getProxy()).build().createRequestFactory();

            GoogleHttpConnector.httpFactoryMap.put(key, requestFactory);
        }

        return requestFactory;
    }

    @Override
    public GoogleResponse getContent(String rqUrl, HttpHostExt proxy, boolean withTimeout) throws IOException {
        try {
            GenericUrl url = new GenericUrl(URI.create(rqUrl));

            if ((GBDOptions.secureMode() && proxy.isLocal()) || !proxy.isAvailable()) return null;

            HttpResponse resp;
            if (this.validateProxy(rqUrl, proxy)) {
                HttpRequest req = this.getFactory(proxy).buildGetRequest(url)
                        .setConnectTimeout(withTimeout ? HttpConnector.CONNECT_TIMEOUT : HttpConnector.CONNECT_TIMEOUT * 10)
                        .setSuppressUserAgentSuffix(true);
                if (this.needHeaders(rqUrl))
                    req.setHeaders(proxy.getHeaders(this.getUrlType(rqUrl)));

                resp = this.getContent(req, proxy, 0);
            } else throw new RuntimeException("Invalid proxy config!");

            if (null == resp)
                HttpConnector.logger.finest(String.format("No response at url %s with proxy %s", rqUrl, proxy.toString()));

            return new GoogleResponse(resp);
        } catch (HttpResponseException hre) {
            HttpConnector.logger.severe("Connection error: " + hre.getStatusMessage());
            throw new GoogleResponseException(hre);
        }
    }

    private HttpResponse getContent(HttpRequest req, HttpHostExt proxy, int attempt) throws IOException {
        if (HttpConnector.MAX_RETRY_COUNT <= attempt) {
            proxy.registerFailure();
            return null;
        }

        if (1 < attempt) try {
            HttpConnector.logger.finest(String.format("Attempt %d with %s url", attempt, req.getUrl().toString()));
            Thread.sleep((long) (HttpConnector.SLEEP_TIME * attempt));
        } catch (InterruptedException ignored) {
        }

        try {
            return req.execute();
        } catch (SocketTimeoutException ste1) {
            proxy.registerFailure();
            return this.getContent(req, proxy, ++attempt);
        }
    }

    @Override
    public void close() {

    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/connectors/google/GoogleHttpConnector.kt
package ru.kmorozov.gbd.core.logic.connectors.google

import com.google.api.client.http.GenericUrl
import com.google.api.client.http.HttpRequest
import com.google.api.client.http.HttpRequestFactory
import com.google.api.client.http.HttpResponseException
import com.google.api.client.http.javanet.NetHttpTransport.Builder
import ru.kmorozov.gbd.core.config.GBDOptions
import ru.kmorozov.gbd.core.logic.connectors.HttpConnector
import ru.kmorozov.gbd.core.logic.connectors.Response
import ru.kmorozov.gbd.core.logic.connectors.Response.Companion.EMPTY_RESPONSE
import ru.kmorozov.gbd.core.logic.proxy.HttpHostExt
import ru.kmorozov.gbd.core.logic.proxy.UrlType
import ru.kmorozov.gbd.logger.Logger
import java.net.SocketTimeoutException
import java.net.URI
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by km on 17.05.2016.
 */
class GoogleHttpConnector : HttpConnector() {

    private fun getFactory(proxy: HttpHostExt): HttpRequestFactory {
        val key = getProxyKey(proxy)

        return httpFactoryMap.getOrPut(key) { Builder().setProxy(proxy.proxy).build().createRequestFactory() }
    }

    override fun getContent(rqUrl: String, proxy: HttpHostExt, withTimeout: Boolean): Response {
        try {
            val url = GenericUrl(URI.create(rqUrl))

            if (GBDOptions.secureMode && proxy.isLocal || !proxy.isAvailable) return EMPTY_RESPONSE

            val resp: Response
            if (validateProxy(rqUrl, proxy)) {
                val req = getFactory(proxy).buildGetRequest(url)
                    .setConnectTimeout(if (withTimeout) CONNECT_TIMEOUT else CONNECT_TIMEOUT * 10)
                    .setSuppressUserAgentSuffix(true)
                if (needHeaders(rqUrl))
                    req.headers = proxy.getHeaders(getUrlType(rqUrl))

                resp = getContent(req, proxy, 0)
            } else {
                logger.error("Invalid proxy config! $proxy")
                proxy.registerFailure()
                return EMPTY_RESPONSE
            }

            if (resp.empty)
                logger.finest("No response at url $rqUrl with proxy $proxy")

            return resp
        } catch (hre: HttpResponseException) {
            if (GBDOptions.debugEnabled)
                logger.info(hre.stackTraceToString())

            if (hre.statusCode != 404)
                logger.severe("Connection error: ${hre.statusMessage}")
            return EMPTY_RESPONSE
        }

    }

    private fun getContent(req: HttpRequest, proxy: HttpHostExt, attempt: Int): Response {
        if (MAX_RETRY_COUNT <= attempt) {
            proxy.registerFailure()
            return EMPTY_RESPONSE
        }

        if (1 < attempt)
            try {
                logger.finest(String.format("Attempt %d with %s url", attempt, req.url.toString()))
                Thread.sleep((SLEEP_TIME * attempt).toLong())
            } catch (ignored: InterruptedException) {
            }

        try {
            return GoogleResponse(req.execute())
        } catch (ste1: SocketTimeoutException) {
            if (GBDOptions.debugEnabled)
                logger.info(ste1.stackTraceToString())

            proxy.registerFailure()
            return getContent(req, proxy, attempt + 1)
        } catch (hre: HttpResponseException) {
            if (GBDOptions.debugEnabled)
                logger.info(hre.stackTraceToString())

            proxy.registerFailure()
            return when (hre.statusCode) {
                403 -> {
                    proxy.reset()
                    req.headers = proxy.getHeaders(UrlType.GOOGLE_BOOKS)
                    getContent(req, proxy, attempt + 1)
                }
                404 -> EMPTY_RESPONSE
                else -> throw (hre)
            }
        }
    }

    override fun close() {

    }

    companion object {

        private val logger = Logger.getLogger(GBDOptions.debugEnabled, GoogleHttpConnector::class.java)
        private val httpFactoryMap = ConcurrentHashMap<String, HttpRequestFactory>()
    }
}
